### PR TITLE
use an in-memory local cache for storing pipeline snapshots

### PR DESCRIPTION
**Pull request by prha**
## Summary
Minimizes deserialization calls to the pipeline snapshot blob for a given snapshot id.

Right now this is done at the SqlRunStorage layer... making sure platform folks see this in case there's a better place to cache that might avoid extra network hops that we wouldn't incur in OSS.

This improves the RunRootQuery time between cold cache and warm cache by 20% for a large pipeline benchmark run.

## Test Plan
BK, loaded single Run view, saw that refreshed Runs had different query times than the initial page load.

Also, saw that the Runs view without https://github.com/dagster-io/dagster/pull/5621 applied had faster query times.

### CODE REVIEW

**Comment by gibsondan** [comment]
this also will not help for OSS users running dagit replicas right? (which is probably not super common, and has other issues too, but just noting)

**Comment by prha** [comment]
Hmm... I think it would still help, but you could potentially see cache misses if you hit different replicas consecutively...

**Comment by prha** [comment]
the plethora of test failures was unexpected... withdrawing from consideration while I figure out what is going on...

**Comment by prha** [comment]
Turns out `functools.lru_cache` was not suitable because it doesn't allow for single cache-key invalidation, caches `None` values, doesn't work on class methods (out of the box) without leaking memory.

I considered something (and wrote an early implementation) that was a lot more magical involving decorators and cache namespaces and multiple inheritance, and opted for this just to avoid magical code.  As we add more and more places where we want to do in-memory caching, might want to resurrect this.

Also, not sure what the maxsize should be for snapshots.  they could be quite large, and holding a bunch in memory might bloat the dagit mem requirements.

File: python_modules/dagster/dagster/core/storage/runs/sql_run_storage.py [hunk 1]
```diff
@@ -54,6 +54,9 @@ class SnapshotType(Enum):
 class SqlRunStorage(RunStorage):  # pylint: disable=no-init
     """Base class for SQL based run storages"""
 
+    def __init__(self):
+        self._snapshot_cache = LRUCache(capacity=32)
```
Comment (dpeng817) [inline python_modules/dagster/dagster/core/storage/runs/sql_run_storage.py:58]: any particular reason why 32

File: python_modules/dagster/dagster/core/storage/runs/sql_run_storage.py [hunk 2]
```diff
@@ -767,6 +784,8 @@ def wipe(self):
             conn.execute(RunTagsTable.delete())  # pylint: disable=no-value-for-parameter
             conn.execute(SnapshotsTable.delete())  # pylint: disable=no-value-for-parameter
             conn.execute(DaemonHeartbeatsTable.delete())  # pylint: disable=no-value-for-parameter
+            if self.snapshot_cache:
+                self.snapshot_cache.clear_all()
```
Comment (dpeng817) [inline python_modules/dagster/dagster/core/storage/runs/sql_run_storage.py:788]: nit but I feel like we're kind of gating on `snapshot_cache`'s existence everywhere

**Review by dpeng817 — APPROVED** [review]
A few nits but things generally look good here.

**Review by prha — COMMENTED** [review]
going to abandon this... seems a little scary/suspect to keep this in the storage layer

**Pull request closed without merging**